### ISSUE TITLE
WEBRTC播放时，音频有时有噪音

### ISSUE THREAD

**Issue by nowgoing** [ossrs/srs#1975]
**描述(Description)**

> 同一个视频推流源，WEBRTC播放时，音频有时有噪音， 但是rtmp拉流就没有噪音

1. SRS版本(Version): `4.0.37`
1. SRS的日志如下(Log):
```
```
1. SRS的配置如下(Config):
```
使用配置文件conf/rtc.conf
```

**重现(Replay)**

> 重现Bug的步骤(How to replay bug?)

1. `ffmpeg rtmp推流`
2. `webrtc播放，检测声音是否有噪音`
3. `没有噪音，停止推流，重复 步骤1-2-3`

**期望行为(Expect)**
> 期望webrtc播放音频没有噪音

**Comment by nowgoing** [comment]
> 
> 
> **描述(Description)**
> 
> > 同一个视频推流源，WEBRTC播放时，音频有时有噪音， 但是rtmp拉流就没有噪音
> 
>     1. SRS版本(Version): `4.0.37`
> 
>     2. SRS的日志如下(Log):
> 
> 
>     1. SRS的配置如下(Config):
> 
> 
> ```
> 使用配置文件conf/rtc.conf
> ```
> 
> **重现(Replay)**
> 
> > 重现Bug的步骤(How to replay bug?)
> 
>     1. `ffmpeg rtmp推流`
> 
>     2. `webrtc播放，检测声音是否有噪音`
> 
>     3. `没有噪音，停止推流，重复 步骤1-2-3`
> 
> 
> **期望行为(Expect)**
> 
> > 期望webrtc播放音频没有噪音

**Comment by nowgoing** [comment]
在函数  `SrsRtcConnection::do_send_packets `中
```
char* iov_base = new char[kRtpPacketSize];
SrsAutoFreeA(char, iov_base);
```
怀疑是没有初始化造成的；使用`memset`对`iov_base `初始化为0后,
```
memset(iov_base , 0, kRtpPacketSize);
```
多次测试，没有再出现噪音。

**Comment by PieerePi** [comment]
See PR #1969.

`TRANS_BY_GPT3`

**Comment by nowgoing** [comment]
> 
> 
> See PR #1969.

`> Test noise solution, thank you`

`TRANS_BY_GPT3`

**Comment by PieerePi** [comment]
I saw that you said you made some changes and solved something before. Is it like what you said, there was no initialization or something like that?

`TRANS_BY_GPT3`

**Comment by nowgoing** [comment]
> 
> 
> I saw that you mentioned before that you made some changes and resolved something. Was it like you said, without any initialization issues?

> No, now we are using your PR.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
I have also encountered noise, which means not being able to hear normal sounds, but only chaotic noise.
Restart the server and then refresh the page, and the noise will be gone. I don't know where this noise is coming from.

`TRANS_BY_GPT3`

**Comment by PieerePi** [comment]
@winlinvip, please take a look at PR #1969.

`TRANS_BY_GPT3`

**Comment by nowgoing** [comment]
> 
> 
> @winlinvip please take a look at PR #1969. 
Make sure to maintain the markdown structure.

Using OBS (setting the `media source as an mp4 file`) for RTMP streaming, setting the audio for streaming as stereo, and playing the stream with WebRTC; the chance of noise reproduction is very high.

@PieerePi 
In the old code, within the function SrsAudioDecoder::decode...
```
for (int i = 0; i < frame_->nb_samples; i++) {
     if (size + pcm_size * codec_ctx_->channels <= max) {
           memcpy(buf + size,frame_->data[0] + pcm_size*codec_ctx_->channels * i, pcm_size * codec_ctx_->channels);      
           size += pcm_size * codec_ctx_->channels;
      }
}
```
**Copying each sample one by one is meaningful to avoid losing too much sample data when it exceeds the maximum value.**
So I made some changes to the old code, as follows...
```
for (int ch = 0; ch < codec_ctx_->channels; ch++) {
       for (int i = 0; i < frame_->nb_samples; i++) {
               if (size + pcm_size  <= max) {
                    memcpy(buf + size, frame_->data[ch] + pcm_size * i, pcm_size);
                    size += pcm_size;
                }
         }
 }
```
There was also no noise in the test.

`TRANS_BY_GPT3`

**Comment by PieerePi** [comment]
Copying each sample one by one is meaningful to avoid discarding too much sample data when it exceeds the maximum limit.

@nowgoing 
If it were me, I would choose to discard the entire channel of data. :-)

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
👍

**Comment by winlinvip** [comment]
Fixed in 4.0.81 https://github.com/ossrs/srs/issues/2106#issuecomment-789688450
